Mathlib/Tactic/DepRewrite.lean

```diff
@@ -482,13 +482,16 @@ def _root_.Lean.MVarId.depRewrite (mvarId : MVarId) (e : Expr) (heq : Expr)
 
 /--
 The configuration used by `rw!` to call `dsimp`.
-This configuration uses only iota reduction (recursor application) to simplify terms.
+This configuration uses just enough simplification to be able to compute
+away applications of `Eq.rec` that `rw!` creates.
 -/
 private def depRwContext : MetaM Simp.Context :=
   Simp.mkContext
     {Lean.Meta.Simp.neutralConfig with
      etaStruct := .none
      iota := true
+     beta := true
+     proj := true
```

Comment:
(We discussed this offline.) Conclusion: running `dsimp` with a strong config may simplify additional terms that `rw!` should not touch. Instead, using the `mdata`-based approach and a custom `dsimproc` would be more self-contained.